### PR TITLE
istioctl: support ip matcher in match tree

### PR DESCRIPTION
**Pull request by howardjohn**
Example:
```
LISTENER                          CHAIN                                                          MATCH                        DESTINATION
envoy://inbound_CONNECT_originate                                                                ALL                          Cluster: inbound_CONNECT_originate
envoy://internal                  inbound-vip|443||echo.default.svc.cluster.local-tcp            ip=10.96.238.42 -> port=443  Cluster: inbound-vip|443|tcp|echo.default.svc.cluster.local
envoy://internal                  inbound-vip|9090||echo.default.svc.cluster.local-tcp           ip=10.96.238.42 -> port=9090 Cluster: inbound-vip|9090|tcp|echo.default.svc.cluster.local
envoy://internal                  inbound-vip|9091||echo.default.svc.cluster.local-tcp           ip=10.96.238.42 -> port=9091 Cluster: inbound-vip|9091|tcp|echo.default.svc.cluster.local
envoy://internal                  inbound-vip|7070||echo.default.svc.cluster.local-http          ip=10.96.238.42 -> port=7070 Cluster: inbound-vip|7070|http|echo.default.svc.cluster.local
envoy://internal                  inbound-vip|443||kubernetes.default.svc.cluster.local-tcp      ip=10.96.0.1 -> port=443     Cluster: inbound-vip|443|tcp|kubernetes.default.svc.cluster.local
envoy://internal                  inbound-vip|9087||shell.default.svc.cluster.local-http         ip=10.96.209.72 -> port=9087 Cluster: inbound-vip|9087|http|shell.default.svc.cluster.local
envoy://internal                  inbound-vip|80||echo-instance.default.svc.cluster.local-http   ip=10.96.57.190 -> port=80   Cluster: inbound-vip|80|http|echo-instance.default.svc.cluster.local
envoy://internal                  inbound-vip|443||echo-instance.default.svc.cluster.local-tcp   ip=10.96.57.190 -> port=443  Cluster: inbound-vip|443|tcp|echo-instance.default.svc.cluster.local
envoy://internal                  inbound-vip|9090||echo-instance.default.svc.cluster.local-tcp  ip=10.96.57.190 -> port=9090 Cluster: inbound-vip|9090|tcp|echo-instance.default.svc.cluster.local
envoy://internal                  inbound-vip|7070||echo-instance.default.svc.cluster.local-http ip=10.96.57.190 -> port=7070 Cluster: inbound-vip|7070|http|echo-instance.default.svc.cluster.local
envoy://internal                  direct                                                         ip=10.244.0.14               Cluster: encap
envoy://internal                  direct                                                         ip=10.244.0.13               Cluster: encap
envoy://internal                  direct                                                         ip=10.244.0.15               Cluster: encap
envoy://internal                  direct                                                         ip=10.244.0.16               Cluster: encap
envoy://outbound-tunnel                                                                          ALL                          Cluster: outbound-tunnel
envoy://inbound_CONNECT_terminate inbound_CONNECT_terminate                                      ALL                          Inline Route:
envoy://                                                                                         ALL                          Inline Route: /healthz/ready*
envoy://                                                                                         ALL                          Inline Route: /stats/prometheus*
```

Needed for https://github.com/istio/istio/pull/42961

fyi @kyessenov

### CODE REVIEW

**Comment by kyessenov** [comment]
I wouldn't say it's needed. This tool has a lot of things missing from upstream Envoy (there are proprietary matchers and inputs).

**Comment by howardjohn** [comment]
Currently it panics, so we should either scrap the idea of having istioctl support for waypoints or implement this IMO. We don't need fully support necessarily, just the ones we use

File: istioctl/pkg/writer/envoy/configdump/listener.go [hunk 1]
```diff
@@ -215,26 +220,44 @@ func recurse(name string, match *matcher.Matcher) (string, bool) {
...
 		equality = "^"
 	case *matcher.Matcher_MatcherTree_CustomMatch:
-		panic("unhandled")
-
+		tc := v.CustomMatch.GetTypedConfig()
+		switch tc.TypeUrl {
+		case IPMatcher:
+			ip := protoconv.SilentlyUnmarshalAny[matcher.IPMatcher](tc)
+			m = map[string]*matcher.Matcher_OnMatch{}
+			for _, rm := range ip.RangeMatchers {
```
Comment (ymesika) [inline istioctl/pkg/writer/envoy/configdump/listener.go:228]: ip can be nil here because of the silent conversion

**Review by ymesika — APPROVED** [review]

**Review by dhawton — APPROVED** [review]
LGTM

**Pull request merged**